src/util/Histogram.h

```diff
@@ -0,0 +1,163 @@
+#include <mutex>
+#include <vector>
+
+// Third party headers
+#include <nlohmann/json.hpp>
+
+namespace lsst::qserv::util {
+
+/// This class is used to help track a value over time.
+/// The `getJson()` function returns a json object that has a structure similar to this
+///	{"HistogramId":"RunningTaskTimes",
+///	 "avg":0.002177499999999979,
+///	 "buckets":
+///	   [{"count":2,"index":0,"maxVal":0.1},
+///		{"count":0,"index":1,"maxVal":1.0},
+///		{"count":0,"index":2,"maxVal":10.0},
+///		{"count":0,"index":3,"maxVal":100.0},
+///		{"count":0,"index":4,"maxVal":200.0},
+///		{"count":0,"index":5,"maxVal":1.7976931348623157e+308}
+///	   ],
+///	 "size":2,
+///	 "total":0.004354999999999958
+///	}
+/// `HistogramId` identifies what the Histogram instance is tracking.
+/// `avg` is the average value of all entries.
+/// `buckets`, each bucket contains the `count` of entries with values <= `maxVal` and > the maxVal of the
+/// previous bucket.
+///    Where the `index` indicates the buckets position in the series.
+/// `size` is the number of entries.
+/// `total` is the sum of all entries.
+/// The `Histogram` keeps a list of all entries. When there are too many entries, the oldest ones are removed
+/// until there are at most `_maxSize` entries. Entries older than `_maxAge` are also removed.
+class Histogram {
+public:
+    using Ptr = std::shared_ptr<Histogram>;
+    using CLOCK = std::chrono::system_clock;
+    using TIMEPOINT = std::chrono::time_point<CLOCK>;
+
+    /// This keeps track of the count of entries with values greater than
+    /// the previous bucket's _maxVal and less than this bucket's _maxVal.
+    class Bucket {
+    public:
+        Bucket(double maxV) : _maxVal(maxV) {}
+        Bucket() = delete;
+        Bucket(Bucket const&) = default;
+
+        double getMaxVal() const { return _maxVal; }
+        int64_t count = 0;
+
+    private:
+        double const _maxVal;
+    };
+
+    class Entry {
+    public:
+        Entry(TIMEPOINT stamp_, double val_) : stamp(stamp_), val(val_) {}
+
+        TIMEPOINT const stamp;  ///< Time of the entry.
+        double const val;       ///< value of the entry.
+    };
+
+    /// Constructor
+    /// @param label - Name for the histogram
+    /// @param bucketVals - a vector indicating the maximum value for each bucket.
+    /// @param maxSize - maximum number of entries to keep. Must be greater than 0.
+    /// @param maxAge - maximum age in milliseconds, such that entries older than
+    ///                       it will be removed from the histogram. Default is 1 hour.
+    Histogram(std::string const& label, std::vector<double> const& bucketVals,
+              std::chrono::milliseconds maxAge, size_t maxSize = 1000);
+    Histogram() = delete;
+    Histogram(Histogram const&) = delete;
+    Histogram& operator=(Histogram const&) = delete;
+
+    /// Add a time and value to the histogram.
+    std::string addEntry(TIMEPOINT stamp, double val, std::string const& note = "");
```

Comment:
The first comment on the method is about its return value. If the return value is still needed by the rest of the code then it's better to be documented here (in the header). I have looked at the PR, and I'm not seeing any use for the method's return, not even in the unit test.

The second comment is that the default value for the string should be `std::string()`:
```
    std::string addEntry(TIMEPOINT stamp, double val, std::string const& note = std::string());
```